sudo: Rework it for the local machine

mvollmer opened the pull request:
Technical design: https://github.com/cockpit-project/cockpit/wiki/Local-sudo,-technical-design

Basic demo: https://www.youtube.com/watch?v=9EPOjg5ugC4
Old multi machine demo: https://www.youtube.com/watch?v=sJ0JhEykvX8

## Release notes

Cockpit now allows more control over your access level.  You can gain
administrative access in a running session, and Cockpit will remember
that for the next session.

Cockpit respects the "sudo" configuration for your account.  For
example, when `NOPASSWD` is in your configuration, you don't have to
provide a password in Cockpit to gain administrative access, and if
"sudo" is configured for two factor authentication, Cockpit will ask
twice.

Cockpit only supports "sudo" to elevate the access level.  If your
custom Cockpit page relies on Polkit to grant access, you now need to
use the `superuser` option for the relevant channels.  For example,
using the UDisks2 D-Bus API now has to look like this:

```
    conn = cockpit.dbus("org.freedesktop.UDisks2", { superuser: "try" })
```

We plan to bring Polkit support back shortly, in enhanced form.

-------------------

FOLLOWUP

- Don't let sudo ask three times before showing an error.
- Implement enough of the old "hint" and "logout" message semantics that restore behavior when using a new cockpit-ws as a bastion host for old cockpit-bridges.
- SSH might get the password that was only remembered for sudo. Keep those separate.
- Update all pages that use `cockpit.permission`. Use `lib/superuser`. Controls should be hidden instead of disabled. No tooltips re permissions.

andreasn commented:
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/d61ff31d5631737b47a94a5cddb1afe72b27c4ce/navigation/admin-access.png)

@garrett I know we discussed some changes to these designs, but I just didn't have time to look into that yet.

mvollmer commented on src/bridge/cockpitpeer.c:
```diff
@@ -860,19 +881,32 @@ cockpit_peer_handle (CockpitPeer *self,
       if (!pipe)
         {
           self->closed = TRUE;
+          startup_done (self, "spawn failed");
```
Can we get a better problem code here?

mvollmer commented on src/bridge/cockpitrouter.c:
```diff
@@ -1219,3 +1311,342 @@ cockpit_router_dump_rules (CockpitRouter *self)
+              if (g_str_equal (id, rule_id))
+                {
+                  g_free (rule_id);
+                  g_info ("starting %s", id);
```
Either remove this, or turn it into a proper journal message

martinpitt commented on src/bridge/cockpitpackages.c:
```diff
@@ -1310,10 +1310,6 @@ cockpit_packages_get_bridges (CockpitPackages *packages)
-          else if (match == NULL)
-            {
-              g_message ("%s: missing \"match\" field in package manifest", package->name);
-            }
```
Should this at least enforce "match OR privileged", to ensure that a bridge definition still has something that cockpit knows how to activate?

mvollmer replied:
Good idea! I'll make this XOR.

mvollmer replied:
This means we need to fully remove the pkexec bridge config from the manifest, but I guess that is ok.

mvollmer commented on pkg/systemd/overview.jsx:
```diff
@@ -148,10 +155,13 @@ class OverviewPage extends React.Component {
                          <div className="ct-overview-header-subheading" id="system_information_os_text">{cockpit.format(_("running $0"), this.state.hostnameData.OperatingSystemPrettyName)}</div>}
                     </div>
                     <div className='ct-overview-header-actions'>
+                        { show_superuser && <SuperuserIndicator key="superuser" /> }
```
"key" is not needed, I think.

mvollmer commented on pkg/lib/superuser.jsx:
```diff
@@ -0,0 +1,88 @@
+/* import { superuser } from "superuser.jsx";
+ *
+ * The "superuser" object indicates whether or not the current page
+ * can open superuser channels or not.
```
s/or not//

mvollmer commented on pkg/lib/superuser.jsx:
```diff
@@ -0,0 +1,88 @@
+ *
+ * This is true when the page can open superuser channels, and false
+ * otherwise.  Right after page load, this field might be "null" until
+ * everything is fully initialized.  During that time, opening a
```
This isn't true.

mvollmer commented on pkg/systemd/overview-cards/configurationCard.jsx:
```diff
@@ -326,42 +326,24 @@ export class ConfigurationCard extends React.Component {
-                             this.permission.user ? this.permission.user.name : '')
-            : this.realmd.button_tooltip);
-        const domain_disabled = !this.permission.allowed || this.realmd.button_disabled;
+        const domain_tooltip = superuser.allowed && this.realmd.button_tooltip;
```
We could move all the superuser handling into realmd.js as well

mvollmer commented on pkg/systemd/overview.jsx:
```diff
@@ -109,34 +106,27 @@ class OverviewPage extends React.Component {
-                    toggle={
-                        <DropdownToggle
+        let headerActions = null;
+        if (superuser.allowed)
```
Use `this.state.privileged` here

martinpitt commented on pkg/systemd/superuser-alert.jsx:
```diff
@@ -0,0 +1,60 @@
+                <Alert hidden={this.superuser.Current != "none" || this.state.dismissed}
+                       variant="info"
+                       action={actions}
+                       title={_("Web console is running in limited access mode.")} />
```
Again, "limited access" is too fuzzy here. "without administrative access"?

martinpitt commented on src/base1/cockpit.js:
```diff
@@ -4328,6 +4328,16 @@ function factory() {
+        function maybe_reload() {
+            if (api.valid && self.allowed !== null) {
+                if (self.allowed != (api.Current != "none"))
+                    window.location.reload(true);
```
Can this look at the page's manifests, so that pages can declare "I watch out for superuser changes and update myself without reloading"?

mvollmer replied:
My plan was that cockpit.permission will be deprecated completely (it never really did all that it was promising). The way for a page to declare that it doesn't want to be reloaded is by stopping to use cockpit.permission.

martinpitt commented on pkg/shell/manifest.json.in:
```diff
@@ -24,7 +24,7 @@
     "bridges": [
         {
-            "privileged": true,
+            "privileged": false,
```
Does this still have any effect now? Preferring sudo is fine of course (I agree), but in cases where sudo doesn't work, pkexec should still be available as a fallback. We did get bug reports about how to configure polkit for allowing a user to run cockpit-bridge as root. (This of course isn't a real subset of root privileges, but it's a nice picket fence to avoid accidents)

mvollmer replied:
> Does this still have any effect now?

Setting `privileged: false` would essentially comment out this bridge. This doesn't work anymore with the extra "privileged XOR match" validation.

mvollmer commented on pkg/shell/index.html:
```diff
@@ -29,6 +29,7 @@
                 <div class="spinner spinner-md spinner-white">
                 </div>
               </li>
+              <li id="super-user-indicator">
```
Autsch, did I introduce that error in one of the earlier commits?

mvollmer replied:
I did...

mvollmer commented on src/bridge/cockpitpolkitagent.c:
```diff
@@ -321,6 +261,13 @@ cockpit_polkit_agent_initiate_authentication (PolkitAgentListener *listener,
 
   result = g_task_new (G_OBJECT (self), NULL, callback, user_data);
 
+  if (!g_str_equal (action_id, "org.cockpit-project.cockpit.root-bridge"))
```
> This action isn't defined anywhere.

It's defined in `src/bridge/org.cockpit-project.cockpit-bridge.policy` and corresponds to running "pkexec /usr/bin/cockpit-bridge".

martinpitt replied:
D'oh, silly me -- I even wrote that! I just looked at the PR.

mvollmer commented on src/bridge/cockpitrouter.h:
```diff
@@ -57,6 +60,13 @@ void                cockpit_router_set_bridges                      (CockpitRout
 
+void                cockpit_router_dbus_startup                    (CockpitRouter *self);
+
+void                cockpit_router_prompt_user_password            (CockpitRouter *self,
```
My idea was that we start with a single specifc prompting function and then later add others... that might not have been a good idea.

mvollmer replied:
This is now more generic and allows arbitrary password-like prompts.

mvollmer commented on src/ws/login.html:
```diff
@@ -56,7 +56,7 @@ <h4 id="login-info-message" class="pf-c-alert__title"></h4>
               <div class="checkbox-row">
                 <input type="checkbox" class="form-control" id="authorized-input">
-                <label for="authorized-input" class="checkbox" translate>Reuse my password for privileged tasks</label>
+                <label for="authorized-input" class="checkbox" translate>Reuse my password for remote connections</label>
```
It's still on by default.  I didn't want to touch anything related to remote connections in this PR.  The checkbox always had the function "Reuse my password for privileged tasks and connections to remote machines." but didn't fully spell it out.  Now it's only for remote connections.

martinpitt replied:
Ack, makes sense, thanks!

mvollmer commented on src/ws/login.js:
```diff
@@ -488,12 +488,18 @@
+            var superuser_key = "superuser:" + user + (machine ? ":" + machine : "");
+            var superuser = localStorage.getItem(superuser_key) || "any";
+            localStorage.setItem("superuser:key", superuser_key);
+            localStorage.setItem(superuser_key, superuser);
```
See `pkg/shell/superuser.jsx` for how `superuser:key` is used.  I didn't want to risk getting login and shell out of sync with regard to the key, so I am passing it explicitly. Maybe that is too paranoid.

> Also, what if my user name is "key"?

Good point!

mvollmer commented on pkg/shell/superuser.jsx:
```diff
@@ -0,0 +1,279 @@
+                <StatelessSelect extraClass="form-control"
+                                 selected={state.method}
+                                 onChange={state.change}>
+                    { state.methods.map(m => <SelectEntry key={m} data={m}>{m}</SelectEntry>) }
```
>  Does it preserve the ordering of the `privileged: true` definitions in the manifest?

Yes.

> My gut feeling is that we should default to `sudo`, i. e. swap the order in the manifest.

"sudo" is now the only method, and then the whole selection step is skipped.

mvollmer commented on pkg/shell/superuser.jsx:
```diff
@@ -0,0 +1,279 @@
+                    .then(() => {
+                        const key = window.localStorage.getItem("superuser:key");
+                        if (key)
+                            window.localStorage.setItem(key, "none");
```
No, there will be no `superuser:key` when a old login page is used on the bastion host.

mvollmer commented on pkg/shell/superuser.jsx:
```diff
@@ -0,0 +1,279 @@
+                        {_("Cancel")}
+                    </Button>
+                    <Button bsStyle='primary' onClick={apply}>
+                        {_("Apply")}
```
Yes, the whole "Lock" dialog hasn't been designed yet.

mvollmer commented:
Old unit tests are green, but integration tests all fail due to unexpected messages from sudo.

mvollmer commented:
Useful integration test run: https://github.com/cockpit-project/cockpit/pull/13482/commits/277735a2ecc74f3d9e2050556f8f2b8fc22f9c6e

mvollmer commented:
Useful test results: https://logs.cockpit-project.org/logs/pull-13482-20200302-084541-c9b80408-fedora-31/log.html

mvollmer commented:
Journal when sudo fails with `semanage login`:
```
Mar 05 09:23:41 localhost.localdomain cockpit-bridge[24324]: starting privileged bridge sudo
Mar 05 09:23:41 localhost.localdomain sudo[24403]: pam_unix(sudo:auth): conversation failed
Mar 05 09:23:41 localhost.localdomain sudo[24403]: pam_unix(sudo:auth): auth could not identify password for [admin]
Mar 05 09:23:41 localhost.localdomain audit[24403]: USER_AUTH pid=24403 uid=1000 auid=1000 ses=24 subj=sysadm_u:sysadm_r:sysadm_sudo_t:s0 msg='op=PAM:authentication grantors=? acct="admin" exe="/usr/bin/sudo" hostname=? addr=? terminal=? res=failed'
Mar 05 09:23:41 localhost.localdomain kernel: audit: type=1100 audit(1583418221.950:806): pid=24403 uid=1000 auid=1000 ses=24 subj=sysadm_u:sysadm_r:sysadm_sudo_t:s0 msg='op=PAM:authentication grantors=? acct="admin" exe="/usr/bin/sudo" hostname=? addr=? terminal=? res=failed'
Mar 05 09:23:41 localhost.localdomain audit[24403]: USER_CMD pid=24403 uid=1000 auid=1000 ses=24 subj=sysadm_u:sysadm_r:sysadm_sudo_t:s0 msg='cwd="/run/user/1000" cmd=636F636B7069742D627269646765202D2D70726976696C65676564 exe="/usr/bin/sudo" terminal=? res=failed'
Mar 05 09:23:41 localhost.localdomain kernel: audit: type=1123 audit(1583418221.954:807): pid=24403 uid=1000 auid=1000 ses=24 subj=sysadm_u:sysadm_r:sysadm_sudo_t:s0 msg='cwd="/run/user/1000" cmd=636F636B7069742D627269646765202D2D70726976696C65676564 exe="/usr/bin/sudo" terminal=? res=failed'
Mar 05 09:23:49 localhost.localdomain systemd[1]: systemd-hostnamed.service: Succeeded.
Mar 05 09:23:49 localhost.localdomain audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-hostnamed comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Mar 05 09:23:49 localhost.localdomain kernel: audit: type=1131 audit(1583418229.509:808): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-hostnamed comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
```

Journal messages with `setenforce 0`

```
Mar 05 09:24:39 localhost.localdomain cockpit-bridge[24324]: starting privileged bridge sudo
Mar 05 09:24:39 localhost.localdomain audit[24432]: AVC avc:  denied  { ioctl } for  pid=24432 comm="sudo" path="socket:[99183]" dev="sockfs" ino=99183 ioctlcmd=0x5413 scontext=sysadm_u:sysadm_r:sysadm_sudo_t:s0 tcontext=sysadm_u:sysadm_r:sysadm_t:s0 tclass=unix_stream_socket permissive=1
Mar 05 09:24:39 localhost.localdomain kernel: audit: type=1400 audit(1583418279.975:812): avc:  denied  { ioctl } for  pid=24432 comm="sudo" path="socket:[99183]" dev="sockfs" ino=99183 ioctlcmd=0x5413 scontext=sysadm_u:sysadm_r:sysadm_sudo_t:s0 tcontext=sysadm_u:sysadm_r:sysadm_t:s0 tclass=unix_stream_socket permissive=1
```

mvollmer commented:
Useful test results: https://github.com/cockpit-project/cockpit/pull/13482/commits/f6a87953fbb9e2dbe400634e5cadbeb9b88b3267

mvollmer commented:
@garrett, here are some screenshots of things that aren't covered by the original mockup:

When going from "Administrative access" back to "Limited access":

![image](https://user-images.githubusercontent.com/3228183/76762821-adfec180-679a-11ea-86a7-a24677385bb4.png)

When nothing needed to be entered to gain "Administrative access":

![image](https://user-images.githubusercontent.com/3228183/76762916-d981ac00-679a-11ea-8f54-a24a92e09190.png)

Arbitrary prompts for 2FA:

![image](https://user-images.githubusercontent.com/3228183/76763223-6167b600-679b-11ea-9f64-f2b9fadd7f2d.png)

garrett commented:
Does the page need to reload when getting or giving up admin access?

mvollmer commented:
> Does the page need to reload when getting or giving up admin access?

The Shell and the Overview do not need to be reloaded, but pretty much every other page.  They do that automatically.

mvollmer commented:
@garrett, up-to-date video: https://www.youtube.com/watch?v=9EPOjg5ugC4  Get's a bit rambling in the end... :-)

garrett commented:
@mvollmer: Wow! That's already super-impressive.

Obviously, there are a few rough edges, but I'm happy to help out. Martin has been debugging why VMs don't build properly on my system — he identified the problem and is making a fix.

But it's already better than both what I expected and what we talked about at the F2F. (It doesn't have to sign out to switch. It doesn't even have to fully reload. That's awesome!)

(It looks like it has to do with Silverblue using /var/home/* and having that symlinked to /home/* — Cockpit's build system has _both_ specified right now, and some commands use one and some use the other. So it gets confused and does bad stuff.)

Anyway, once the bug is fixed, I build VMs and should be able to test this out and make some fixes. I'm looking forward to it!

mvollmer commented:
`make check-memory` fails.

garrett commented:
![Screenshot_2020-04-08 Overview - localhost localdomain(1)](https://user-images.githubusercontent.com/10246/78781218-21fa4700-79a0-11ea-9031-8aaed48dc6b2.png)

As the modal button reordering PR has landed (#13546):

1. Button order is (now) incorrect
2. Cancel should be a link-styled button
3. Spinner is on the wrong side of the buttons now (not pictured, as it's _quick_)

![Screenshot_2020-04-08 Overview - localhost localdomain](https://user-images.githubusercontent.com/10246/78781370-6d145a00-79a0-11ea-881a-f36bad3f7005.png)

4. Notifications at the top of the overview are inconsistent
5. There's too much prominence to the notification (the shadow is excessive) — I'm sure this is PF4 related
6. _(Out of scope for this PR)_ Last login probably shouldn't be a dismissable notification anyway, but part of system information (I need to open a new bug on this)
7. When I was talking about a notification at the top, at our DevConf F2F, I was thinking above the system name, flush to the edges (and not dismissable)

![Screenshot_2020-04-08 Overview - localhost localdomain(2)](https://user-images.githubusercontent.com/10246/78781747-15c2b980-79a1-11ea-96cc-42e6d0a8058d.png)

8. The limited / administrative access label does _not_ look clickable at all
   - it doesn't look clickable by default
   - it doesn't change the mouse cursor from an I-beam to a finger (or even a standard pointer)
   - it doesn't react on hover whatsoever

![Screenshot_2020-04-08 Overview - localhost localdomain(3)](https://user-images.githubusercontent.com/10246/78781971-7d790480-79a1-11ea-9fa4-8c56774302d7.png)

9. Similar issues with the other dialog: Incorrect button order (now) and cancel needing to be restyled
10. Text needs work
    - Title suggestions (we can talk about if any of these, or a modification of these may work in IRC — or if there's something even better — we probably also want to adjust the escalation modal to be the opposite of what is chosen here):
       - Drop administration privileges
       - De-escalate privileges
       - Limit access
       - Revoke administration access
       - Switch to limited access mode
       - Lower access level
       - Limit access
    - Body suggestions (may need to be matched with the title):
       1. First paragraph (something to this effect, with adjusted language):
          - Lowering the access level will reduce functionality of the web console.
       2. Second paragraph (something to this effect; perhaps add it to the escalation modal too):
          - Your browser will remember your access level across sessions.
    - Action button suggestions:
       - Drop privileges
       - Switch to limited access mode
       - Limit access
       - (etc.; the verb used for the button should be based on the above text for the title and possibly the body text as well)

martinpitt commented:
> Notifications at the top of the overview are inconsistent
> There's too much prominence to the notification (the shadow is excessive) — I'm sure this is PF4 related

Use the `isInline` attribute with the alerts, that's what we do for all the others. That will solve both issues.

As for the text:

> Drop administration privileges
> Action button suggestions:
>    Drop privileges

I like these. "Limited access" is rather fuzzy, and it's hard to correlate to any Linux concept. If we just can't get along with negative pronouns/verbs of "Administrator privileges", then perhaps "user privileges"? But perhaps we can make-do with just "enable/disable/get/drop/enter/leave" action pairs, and *only* use "administration"/"administrative"?

garrett commented:
> I like these. "Limited access" is rather fuzzy

That's why I was thinking the body text is a good place to tie it in.

We don't want something negative or anti- at the top, hence the "limited access" label. It's limited access to Cockpit / Web Console / the system due to not having administration privs.

The modal is a good place to have a clear message for Windows admins, new-to-Linux admins, and seasoned Linux gurus (either in the title and/or body, depending on the audience).

mvollmer commented:
> "Limited access" is rather fuzzy, 

I think we should start from what we put into the indicator label. The rest follows from there, to keep it consistent.

Right now it's "Administrative access" and "Limited access".

Personally, I am okay with "Limited access".

garrett commented:
OK, here's a start (with the trickier of the two dialogs):

> ### Switch to limited access
> Limited access mode restricts administration privileges. Some parts of the web console will have reduced functionality.
>
> Your browser will remember the access level across sessions.
> [Limit access] [Cancel]

(We could, in the future, make the remembering feature a default option, if we wanted... "☑ Remember limited access level across sessions." But that's out of scope here, and we may not want to add it anyway.)

martinpitt commented:
I tested the current version interactively. First of all, nice work! It feels very natural, great work!

Many of @garrett's points above were addressed, but at least point 8 is still outstanding (access indicator in panel does not look like a button).

Some observations:

 - [ ] The unprivileged Alert looks a bit mis-placed: Shouldn't it be at the same spot and with the same margins as the "Last Login" one?

    ![image](https://user-images.githubusercontent.com/200109/78881917-bb386480-7a57-11ea-98f7-e7bc3eb64b51.png)

- [ ] The "Remember my password for remote machines" check box is enabled by default. Is that an explicit decision, or just a leftover from how we treated the sudo one? I'd be in favor of having it off by default, for compatibility how ssh behaves. Or is that for Dashboard backwards compat? Is there any way we can get rid of that box in the future?

- [ ] When I type a wrong password into the "Administrative access" dialog, I get no feedback the first two times, and the third time it says "This didn't work, try again". That smells like it mirrors the PAM protocol for sudo, and we can't actually tell in between whether the password was wrong or we get a different prompt. We could perhaps check if the prompt is the same as the previous one? But this is minor, no need to block this PR.

- :+1: I tested a few different pages when switching privs; the auto-reload works well, and the "Read-only - Requires administration access to edit" thing on a services page is a nice extra treat!

- [ ] I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. That's a regression. Also, there's no indicator at all in the panel though -- is that by design, or some omission? Should it show "Limited access"?

- [ ] I changed my VM password to match the remote password, and added that machine again. I wasn't asked for the password any more for ssh (→ "Reuse my password for remote systems" worked) but I still don't have admin there. That's also a regression.

- [ ] I logged into the remote system straight from the login page ("bastion host"). I don't have sudo powers. That's again a regression, presumably very similar/the same one. Can we please cover this in a test case?

martinpitt commented:
It looks like test failures [like this](https://logs.cockpit-project.org/logs/pull-13482-20200408-141515-d49b6cee-debian-stable/log.html#137-2) actually spot that "no admin on remote system" regression?

garrett commented:
Oh, this is looking great!

I need to apply some small tweaks (probably just CSS) today.

Meanwhile, I did notice it's really odd how things act when you're signed into an account that _only_ has limited access. There's a banner that teases you with the ability to upgrade your access (which you don't have privileges for) and clicking the limited access button in the header still prompts you for your password to become an admin — which you cannot do.

We need to fix this.

martinpitt commented on pkg/shell/superuser.jsx:
```diff
@@ -288,9 +288,9 @@ export class SuperuserIndicator extends React.Component {
         function create_trigger(unlocked, onclick) {
             return (
-                <span className="navbar-text" onClick={onclick}>
+                <a className="navbar-text" onClick={onclick}>
```
Should we actually turn it into a `Button`? at least `variant="link"`, but maybe even "secondary" so that it looks more clickable?

garrett commented:
> The unprivileged Alert looks a bit mis-placed: Shouldn't it be at the same spot and with the same margins as the "Last Login" one?

No, it was there before. I requested it to be where it is. I need to restyle it, however. I have plans to make it feel better than it does right now.

It shouldn't be dismissable, and it's not really a notification.

martinpitt commented on pkg/systemd/superuser-alert.jsx:
```diff
@@ -33,15 +33,13 @@ export class SuperuserAlert extends React.Component {
         super();
         this.superuser = cockpit.dbus(null, { bus: "internal" }).proxy("cockpit.Superuser", "/superuser");
         this.superuser.addEventListener("changed", () => { this.setState({}) });
-        this.state = { dismissed: false, show_unlock_dialog: false };
```
Why would we not make the alert dismissable? If someone really does not want to see this any more, I don't see a reason to force this message to stay around. You can always use the "Limited" indicator to go back to sudo.

martinpitt commented on src/bridge/cockpitpackages.c:
```diff
@@ -1322,6 +1327,8 @@ cockpit_packages_get_bridges (CockpitPackages *packages)
             {
               result = g_list_prepend (result, item);
             }
+
+
```
Spurious whitespace

mvollmer replied:
Fixed.

mvollmer replied:
Fixed in the wrong commit...

martinpitt commented on src/bridge/cockpitpeer.c:
```diff
@@ -172,6 +176,27 @@ on_timeout_reset (gpointer user_data)
+
+  if (self->startup_auth_cookie)
+    {
+      GBytes *reply = cockpit_transport_build_control ("command", "authorize",
```
Not at all required, just FYI: We recently moved to using `g_autoptr(GBytes) reply` for such stuff, so that we don't need to explicitly unref stuff any more.

martinpitt commented on src/bridge/cockpitpeer.c:
```diff
@@ -262,9 +289,41 @@ on_other_control (CockpitTransport *transport,
+      else if (cockpit_json_get_bool (self->config, "privileged", FALSE, &privileged) && privileged)
+        {
+          if (self->startup_auth_cookie)
+            g_message ("%s: received overlapping \"authorize\" requests", self->name);
+          else if (!self->router)
+            g_message ("%s: no router for answering \"authorize\" request", self->name);
```
Can these happen in practice? If not, these should be `g_warning()`, so that our tests will spot them.

mvollmer replied:
Good point. The former can happen when stuff happens concurrently that shouldn't. The latter can only happen with a custom bridge that doesn't do the initialization correctly. So, both fixed to be warnings.

mvollmer replied:
But using that logic, things like
```
g_message ("%s: received \"authorize\" request without a valid cookie", self->name);
```
should also be a warning, no?

martinpitt commented on src/bridge/cockpitrouter.c:
```diff
@@ -1132,6 +1217,8 @@ cockpit_router_add_bridge (CockpitRouter *self,
   g_return_if_fail (COCKPIT_IS_ROUTER (self));
   g_return_if_fail (config != NULL);
 
+  // TODO - reject dynamic privileged bridges
```
Not sure what that means exactly, but should that block this PR?

mvollmer replied:
Bridges are usually spawned when the first channel is opened that requires them, and the command that is spawned can be constructed from options of that first channel.  The ssh bridges are an example.

This doesn't make sense for superuser bridges: We only want one of them, and there are no channel options when they are started.

mvollmer replied:
Also, the rest of the code assumes that `rule->user_data` is a CockpitPeer.  I have implemented this now.

garrett commented:
Improved the look of of the sudo banner.

![Screenshot_2020-04-09 Overview - localhost localdomain](https://user-images.githubusercontent.com/10246/78895600-6a346a80-7a6f-11ea-8d23-8cbefdf55bb4.png)

garrett commented:
I think I want to make the banner a little darker, perhaps... looking over it again, it might get washed out. We would want to, of course, keep AA contrast. :wink:

Also, we should perhaps get PF4 to support a banner concept as either a new widget or a style of notification (and switch to it if something like this becomes available in the future). But that's outside of this PR. (Just mentioning upstreaming this at some point.)

mvollmer commented:
> It's really odd how things act when you're signed into an account that _only_ has limited access.
> 
>     * There's a banner that teases you with the ability to upgrade your access (which you don't have privileges for) and clicking the limited access button in the header still prompts you for your password to become an admin — which you cannot do.
> 
>     * We probably need to handle this case both by hiding the button in the notifier at the top of the overview page and having a different dialog that pops up when clicking on the header's admin priv button.

Currently, we swallow all the error output of sudo and just say "Sorry, this didn't work.". Would it be enough to show the actual error messages from sudo?  In this case, that would be "Sorry, user admin may not run sudo on dev.", after the user has successfuly authenticated.

mvollmer commented:
>    * We probably need to handle this case both by hiding the button in the notifier at the top of the overview page and having a different dialog that pops up when clicking on the header's admin priv button.

This can be done by running `sudo -v -n` and carefully inspecting the output.  If it matches `"Sorry, user $USER may not run sudo on $HOST." then we can assume that the logged in user can't even run sudo.

This would make our UI code sudo specific, but I think that's ok for now.

mvollmer commented:
> The "Remember my password for remote machines" check box is enabled by default. Is that an explicit decision, or just a leftover from how we treated the sudo one? I'd be in favor of having it off by default, for compatibility how ssh behaves. Or is that for Dashboard backwards compat?

This is indeed for backwards compatibility.

> Is there any way we can get rid of that box in the future?

Yes, that was one of the initial goals: https://github.com/cockpit-project/cockpit/wiki/Sudo--Multi-Machine-Overhaul

I'll return to that once both this and the new navigation has landed.

mvollmer commented:
> When I type a wrong password into the "Administrative access" dialog, I get no feedback the first two times, and the third time it says "This didn't work, try again". That smells like it mirrors the PAM protocol for sudo, and we can't actually tell in between whether the password was wrong or we get a different prompt. 

Yeah, exactly. I am not happy with that either. We probably need to teach "sudo" a new "--prompt-me-only-once" option. Even in the shell (with history), being asked three times is annoying IMO.

> We could perhaps check if the prompt is the same as the previous one? But this is minor, no need to block this PR.

I was thinking the same, hmm.... maybe we do just this.

mvollmer commented:
> * I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. That's a regression. Also, there's no indicator at all in the panel though -- is that by design, or some omission? Should it show "Limited access"?

There should be no regressions like this, not by design.  There should be tests for this, and they should pass. I'll check again.

mvollmer commented:
> There should be tests for this, and they should pass. I'll check again.

Ah, but not for old machines on the dashboard, I'll work on that.

mvollmer commented:
> I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. That's a regression.

Actually, this has never worked. If the users are different, cockpit-ws will not hand out the password.

mvollmer commented:
> It looks like test failures [like this](https://logs.cockpit-project.org/logs/pull-13482-20200408-141515-d49b6cee-debian-stable/log.html#137-2) actually spot that "no admin on remote system" regression?

No, that test was temporarily broken due to the UI polish FIXUPs.

mvollmer commented:
>  -  I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. That's a regression. 

This is the same behavior as without this PR.  It's not ideal, but I don't want to fix it here yet.

>    Also, there's no indicator at all in the panel though -- is that by design, or some omission? Should it show "Limited access"?

That is by design in this PR.  After we have #13663, we need to revisit that.

> -  I changed my VM password to match the remote password, and added that machine again. I wasn't asked for the password any more for ssh (→ "Reuse my password for remote systems" worked) but I still don't have admin there. That's also a regression.

This is also the same behavior as without this PR.  As long as the user names don't match, the remote machine wont get the password during superuser escalation.

There is TestSuperuserDashboard a TestSuperuserDashboardOldMachine to test the behavior in these scenarios.

> -  I logged into the remote system straight from the login page ("bastion host"). I don't have sudo powers. That's again a regression, presumably very similar/the same one. Can we please cover this in a test case?

This should work, and TestSuperuserOldShell tests this.

What does not work is logging into a old machine using a new machine as the bastion host, and _not_ automatically becoming superuser on the remote machine.

martinpitt commented on pkg/shell/superuser.jsx:
```diff
@@ -176,19 +176,19 @@ export class SuperuserDialogs extends React.Component {
-                });
-            }
+            cockpit.spawn(["sudo", "-v", "-n"], { err: "message" })
+                    .fail(e => {
```
This is an ES6 code file, please use standard Promise API, i. e. `.catch()` instead of `.fail()`. Also, I think `.then()` needs to go first, otherwise this may behave weird due to promise chaining. I recently had a case like this where the .then handler received the caught error.

mvollmer replied:
Yes, I had fun debugging this when switching to `.catch(...)`...  Done in the the new `can_do_sudo` function.

martinpitt commented on pkg/shell/superuser.jsx:
```diff
@@ -176,19 +176,19 @@ export class SuperuserDialogs extends React.Component {
-            }
+            cockpit.spawn(["sudo", "-v", "-n"], { err: "message" })
+                    .fail(e => {
+                        if (e.message && e.message.match("Sorry, user .+ may not run sudo on .+\\.")) {
```
I'm a bit sceptical about pinning down human visible strings like this. It may slightly change over time, or even get translated. Perhaps run it with `LC_ALL=C` and just check for "may not run"?

mvollmer replied:
Discussed at `can_do_sudo`.

martinpitt commented on pkg/systemd/superuser-alert.jsx:
```diff
@@ -33,14 +33,27 @@ export class SuperuserAlert extends React.Component {
+                    if (e.message && e.message.match("Sorry, user .+ may not run sudo on .+\\.")) {
+                        this.setState({ can_do_sudo: false });
+                    } else
+                        this.setState({ can_do_sudo: true });
```
This is confusing and needs a comment -- if `sudo -v -n` fails, why is it set to true? Also, same issues from above apply.

martinpitt commented on pkg/shell/superuser.jsx:
```diff
@@ -144,13 +144,13 @@ export class SuperuserDialogs extends React.Component {
         this.superuser.addEventListener("changed", () => {
             this.setState({
-                show: this.superuser.Current != "root",
+                show: this.superuser.Current != "root" && this.superuser.Current != "",
```
So this value can take `"none"`, `"root"`, and what else? Comparing against `!= ""` is not self-explanatory here. Maybe a more standard tristate of null/false/true would be cleaner?

mvollmer replied:
Yes, agreed, this isn't clear enough anymore.

The `""` value is used during startup of the bridge until the value either goes to "sudo" or "none".  During that time, the UI should pretend that we have admin access, but the page should not reload if the value goes to "none".

I guess using ´"init"` instead of `""` might make thigns a bit clearer.  I'll have a think about this.

mvollmer replied:
>  During that time, the UI should pretend that we have admin access

Should it, actually?  I am starting to get confused...

martinpitt replied:
I think it should start with "limited". We used to always enable the privileged operations on the page until sudo detection was complete, and it led/leads to confusion. I had some gut feeling that we fixed that recently. It seems less confusing to start unprivileged until the sudo detection is done (and positive).

mvollmer replied:
Ok, I changed this to use `"init"` and documented the values and possible transitions.

mvollmer replied:
> I think it should start with "limited". We used to always enable the privileged operations on the page until sudo detection was complete, and it led/leads to confusion. 

Yes, indeed.  It is usually very fast to get admin or to figure out that it wont work, so there is no actual difference in the normal case.  However, sometimes sudo asks three times with a delay in between, or the PAM stack uses something that is slow.

I'll see if I can make the "init" phase fail faster.  We only answer the first prompt anyway, and we can kill the whole process when a second prompt comes.

martinpitt commented on pkg/shell/superuser.jsx:
```diff
@@ -27,6 +27,12 @@ import "form-layout.scss";
+export function can_do_sudo() {
+    return cockpit.spawn(["sudo", "-v", "-n"], { err: "out" })
+            .then(() => true,
+                  (err, out) => !(err.exit_status == 1 && out.match("Sorry, user .+ may not run sudo on .+\\.")));
```
Ah, this addresses some issues above (DRY and the wrong fail/then order), but the too precise string matching and locale dependencies still remain. I. e. please run with `LC_ALL=C` and just check for `indexOf("may not run")`.

But the exit comparison seems wrong; it would now return true on a different exit code.

mvollmer replied:
I wanted this to be really specific, since having a false match will falsly prevent people from gaining root.

martinpitt replied:
OK, our tests should pick up string changes. But let's please pin the locale to C at least.

mvollmer replied:
Done.

martinpitt commented on src/bridge/cockpitrouter.c:
```diff
@@ -1631,6 +1665,102 @@ cockpit_router_dbus_startup (CockpitRouter *router)
+  router->superuser_init_in_progress = TRUE;
+  router->superuser_init_next_rule = router->rules;
+  router->superuser_init_id = g_strdup (id);
+  router->superuser_init_password = g_strdup (password);
```
We are rather careless about the password. This of course already applies to the current implementation, but let's try better. We keep the password in cleartext in an unprivileged process, and copy it around *a lot*. As a first step, could superuser_init() not dup it, but pass it to this function as non-const, and `router` takes ownership of it? Let's please also mlock() the password as soon as we get it, so that it doesn't get swapped onto the disk in clear.

mvollmer replied:
Yes... Pasing the password in the "init" message is not so great either.  It used to be in "authorize" messages, but it is arguably worse to suddenly turn the well know "init" message into a carrier of secrets. I have a look at the new memfd infrastructure that we have.

mvollmer replied:
Ok, I have added a WIP commit that uses the old `authorize` messages to transport the password.  That is not great, but at least not worse that what we have been living with all these years.

mvollmer replied:
We go with this approach.

martinpitt commented on src/bridge/cockpitrouter.c:
```diff
@@ -1631,6 +1665,102 @@ cockpit_router_dbus_startup (CockpitRouter *router)
+      router->superuser_init_in_progress = FALSE;
+      g_free (router->superuser_init_id);
+      router->superuser_init_id = NULL;
+      g_free (router->superuser_init_password);
```
Please also overwrite the password before freeing it, so that it doesn't remain in memory. As it's also contained in the init message, the same mlock+clear+zero copy should be done to it.

martinpitt commented on src/ws/cockpitwebservice.c:
```diff
@@ -598,14 +611,60 @@ process_transport_init (CockpitWebService *self,
+              password_bytes = g_bytes_get_data (password, &password_len);
+              password_hex = cockpit_hex_encode (password_bytes, password_len);
+              json_object_set_string_member (superuser_options, "password", password_hex);
+              g_free (password_hex);
```
Again, please clear the password here. This is short-lived, so presumably we don't need mlock().

martinpitt commented on test/common/testlib.py:
```diff
@@ -528,6 +528,8 @@ def try_login(self, user, password, authorized=True):
         self.set_val('#login-user-input', user)
         self.set_val('#login-password-input', password)
         self.set_checked('#authorized-input', authorized)
```
Should this then be only done for "distropkg"?

mvollmer replied:
Hmm, good point.  The "#authorized-input" now only affects SSH, and we could make that explicit in the tests.

My plan is to eventually remove `#authorized-input` completely. Would that be a better time to figure this out?

martinpitt replied:
We can clean it up in a follow-up for sure. I'd just like to have at least one explicit test case that makes sure that sudo works with that box unchecked.

mvollmer replied:
Indeed, that is a very good point!  I'll add that.

mvollmer replied:
Ok, I did this more properly after all.  login_and_go (et al) now have a "superuser" parameter (that is completely independent from "authorized") and the tests use that insted of "authorized". This means whatever we do about the #authorized-input should not affect most tests in the future.

I have also added a test with `login_and_go(authorized=False, superuser=True)`.

martinpitt commented:
> > I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. That's a regression.
> 
> Actually, this has never worked. If the users are different, cockpit-ws will not hand out the password.

Indeed, I tested this again, and I must have been confused last time. The actual administrative access behaves exactly the same as on master, just the privilege indicator has a regression. I updated the TODO list above.

mvollmer commented:
> * I added a machine with Cockpit 216 to the dashboard (different user/password), and have no admin there. 216 behaves the same way, so this is _not_ a regression. However, there's no indicator at all in the panel though -- is that by design, or some omission? Should it show "Limited access"?
> * In my VM I created a user/password that matches my remote account, and added that machine again. I wasn't asked for the password any more for ssh (→ "Reuse my password for remote systems" worked). I do have sudo powers (on Overview, Storage, etc), but the top panel again does not show anything (should show "Administrative access"). That is a regression, with 216 I get the "Privileged" indicator. Can we please cover this in a test case?

That is by design for this PR: The superuser indicator in the shell is only for the local machine.

mvollmer commented:
> * I logged into the remote system straight from the login page ("bastion host"). Same behaviour as for the previous point -- I do have admin, but indicator in the top panel is empty.

This was also by design: The old indicator corresponded directly to the "[ ] Reuse my password" checkbox, not to whether or not the page actually has admin privs.  I found that indicator and the machinery to be confusing and decided to get rid of the whole shebang.

This means that there is no way anymore to drop admin privs when using a new cockpit-ws as the bastion for a old cockpit-bridge.


We could try to bring the old behavior back, by carefully sending out the old "hint" messages and implementing the old "logout" message semantics. Looking back now, I think that might actually be worthwhile.

garrett commented:
@mvollmer: Anything specific I should look at?

I've filed a few tangential bugs that impact the experience a bit:
- #13942 Re-add space between paragraphs
- #13882 Last login time's display is odd
- #13941 Strip whitespace before and after MOTD
- #13490 Drop MOTD margin, use grid gap instead (already merged)

All of these are unrelated, but do affect the overall Overview experience a bit, which can (in some cases) detract from sudo.

One of the above issues (the paragraph issue) does directly impact does affect the display of the de-escalation dialog. With the fix, it looks like this:

![Screenshot_2020-04-22 Overview - debian(1)](https://user-images.githubusercontent.com/10246/79985606-21d36e80-84ab-11ea-8255-e3bbd97a4f09.png)

mvollmer commented on src/bridge/cockpitpeer.c:
```diff
@@ -74,16 +74,32 @@ struct _CockpitPeer {
   PROP_0,
   PROP_TRANSPORT,
-  PROP_CONFIG
+  PROP_ROUTER,
```
This change is in the wrong commit.

mvollmer replied:
Fixed

mvollmer commented on test/common/testlib.py:
```diff
@@ -528,6 +528,8 @@ def try_login(self, user, password, authorized=True):
         self.set_val('#login-password-input', password)
         self.set_checked('#authorized-input', authorized)
+        if authorized is not None:
+            self.eval_js('window.localStorage.setItem("superuser:%s", "%s");' % (user, "any" if authorized else "none"))
```
This should go to the big test fixing commit.

mvollmer replied:
Done.

mvollmer commented on test/common/testlib.py:
```diff
@@ -574,6 +576,7 @@ def relogin(self, path=None, user=None, authorized=None):
         self.set_val("#login-password-input", self.password)
         if authorized is not None:
             self.set_checked('#authorized-input', authorized)
+            self.eval_js('window.localStorage.setItem("superuser:%s", "%s");' % (user, "any" if authorized else "none"))
```
This should go to the big test fixing commit.

mvollmer replied:
Done

martinpitt commented on test/verify/check-superuser:
```diff
@@ -0,0 +1,429 @@
+import parent
+from testlib import *
+
+class TestSuperuser(MachineCase):
```
Supposedly all classes here need to be skipped on rhel-8-2-distropkg?

mvollmer commented on test/common/testlib.py:
```diff
@@ -565,7 +567,7 @@ def logout(self):
         self.expect_load()
 
-    def relogin(self, path=None, user=None, authorized=None):
+    def relogin(self, path=None, user=None, authorized=None, superuser=None):
```
It's the same for `authorized`, and I copied that. But, yes, it's not super well thought out.

martinpitt replied:
I actually understand that better after reading the new tests. This doesn't really "disable" superuser, it just means "keep the previous state", so it does make sense for relogin(). So its meaning is more like "explicitly change/set superuser"

mvollmer commented:
> It now also fails all the storage tests -- that's actually the kind of protocol incompatiblity for which we have these -distropkg tests in the first place. 

Yes... The storage page doesn't show the privileged buttons.  I think we are mixing a old cockpit-bridge and a new storage page.  This surprises me every time... but we do need to support that, right?

martinpitt commented:
@mvollmer: Yes, we do I'm afraid. That's the kind of mixing that happens in RHEL all the time. cockpit-storage is in Extras (RHEL7)/AppStream (RHEL 8), and bridge is in Base(OS).

martinpitt commented:
The FIXUP looks good to me at first sight, there were just a lot of unrelated test failures. Retrying.

martinpitt commented on test/common/testlib.py:
```diff
@@ -912,8 +912,13 @@ def tearDown(self):
         shutil.rmtree(self.tmpdir)
 
-    def login_and_go(self, path=None, user=None, host=None, authorized=True, superuser=True, urlroot=None, tls=False):
+    def login_and_go(self, path=None, user=None, host=None, authorized="unset", superuser=True, urlroot=None, tls=False):
```
Annoying nitpick from the peanut gallery: The canonical value for "unset" is `None` in Python.

mvollmer replied:
Yeah, but `login_and_go(authorized=None, superuser=...)` needs to mean something different than `login_and_go(superuser=...)`... This sure is ugly; I try to come up with something better.

martinpitt commented:
There's one remaining `testTroubleshooting` failure that happens on most OSes: e. g. [fedora-31](https://logs.cockpit-project.org/logs/pull-13482-20200428-054932-f039ca20-fedora-31/log.html#218-2) or [rhel-8-2](https://logs.cockpit-project.org/logs/pull-13482-20200428-054932-f039ca20-rhel-8-2/log.html). [TestShutdownRestart on fedora-coreos](https://logs.cockpit-project.org/logs/pull-13482-20200428-054932-f039ca20-fedora-coreos/log.html#253-2) also failed, which doesn't ring a bell (i. e. it's not usually flaky). The other failures seem "regular flakes" to me.

mvollmer commented:
> There's one remaining testTroubleshooting failure that happens on most OSes.

Yes, that's not a flake, I have a look.

martinpitt commented:
@mvollmer: testTroubleshooting was an unfortunate fallout of the recent fedora-31 refresh, fixed by #13988. I'll retry the tests.

martinpitt commented:
[TestShutdownRestart on fedora-coreos](https://logs.cockpit-project.org/logs/pull-13991-20200429-120911-c6b6c740-fedora-coreos/log.html#204-2) also fails on other PRs, so that seems unrelated after all.

mvollmer commented:
> > There's one remaining testTroubleshooting failure that happens on most OSes.
> 
> Yes, that's not a flake, I have a look.

Right, the test tried to login without "Reuse my password for remote connections", but SSH got the password anyway via the superuser startup machinery.

Fixing the test is straightforward, but now I am thinking that maybe we need to already now separate the two password uses much more clearly...

mvollmer commented:
> @mvollmer: testTroubleshooting was an unfortunate fallout of the recent fedora-31 refresh, fixed by #13988. I'll retry the tests.

Hmm, I don't actually think so...

mvollmer commented:
> > @mvollmer: testTroubleshooting was an unfortunate fallout of the recent fedora-31 refresh, fixed by #13988. I'll retry the tests.
> 
> Hmm, I don't actually think so...

Right, are you maybe confusing TestMultiMachine.testTroubleshoot and check-setroubleshoot?

martinpitt commented:
@mvollmer: Spot-on; I have trouble with telling apart the troubleshooters, clearly :)

martinpitt commented:
Just some common machines flakes left, nice!

mvollmer commented:
> Just some common machines flakes left, nice!

The multi-machine failures on fedora-31/firefox look a bit suspicious... let's retry.

marusak commented:
I guess this could wrapper better? :)
![Screen Shot 2020-05-06 at 14 40 46](https://user-images.githubusercontent.com/12330670/81177985-cbd0f180-8fa7-11ea-9e50-5667b6dd6737.png)

andreasn commented:
> I guess this could wrapper better? :)

Separate issue, please!

garrett commented:
>>    I guess this could wrapper better? :)

Yes.

> Separate issue, please!

Agreed.

I think this is a PF4-related issue, actually. We can easily fix it, but it should be shoved upstream too. (But we'll work on it in another PR.)
